provider/packet/environ.go

```diff
@@ -0,0 +1,561 @@
+	if len(inst.Plan.Specs.Cpus) > 0 {
+		cpus = uint64(inst.Plan.Specs.Cpus[0].Count)
+	}
+
+	hc := &instance.HardwareCharacteristics{
+		Arch: &amd64,
+		Mem:  &mem,
+		// RootDisk: &instanceSpec.InstanceType.RootDisk,
+		CpuCores: &cpus,
+	}
+
+	return &environs.StartInstanceResult{
+		Instance: inst,
+		Hardware: hc,
+	}, nil
+}
+
+// supportedInstanceTypes returns the instance types supported by packet.
+func (e *environ) supportedInstanceTypes() ([]instances.InstanceType, error) {
+	plans, _, err := e.packetClient.Plans.List(new(packngo.ListOptions))
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported instance types")
+	}
+
+	var instTypes []instances.InstanceType
+nextPlan:
+	for _, plan := range plans {
+		if !validPlan(plan) {
+			continue
+		}
+
+		// TODO: Only work with this plan type for now to avoid
+		// spinning up more expensive instances while testing.
+		if plan.Name != "t1.small.x86" {
+			continue
+		}
+
+		var instArch string
+		switch {
+		case strings.HasSuffix(plan.Name, ".x86"):
+			instArch = arch.AMD64
+		case strings.HasSuffix(plan.Name, ".arm"):
+			instArch = arch.ARM64
+		default:
+			continue nextPlan
+
+		}
+
+		mem, err := parseMemValue(plan.Specs.Memory.Total)
+		if err != nil {
+			continue
+		}
+
+		instTypes = append(instTypes,
+			instances.InstanceType{
+				Id:       plan.ID,
+				Name:     plan.Name,
+				CpuCores: uint64(plan.Specs.Cpus[0].Count),
+				Mem:      mem,
+				Arches:   []string{instArch},
+				// Scale per hour costs so they can be represented as an integer for sorting purposes.
+				Cost: uint64(plan.Pricing.Hour * 1000.0),
+				// TODO: returned by packet's API but not exposed by the packngo client
+				// Deprecated: plan.Legacy,
+			})
+	}
+
+	return instTypes, nil
+}
+
+func validPlan(plan packngo.Plan) bool {
+	isInvalid := plan.Pricing == nil ||
+		plan.Specs == nil ||
+		plan.Specs.Memory == nil ||
+		len(plan.Specs.Cpus) == 0 || plan.Specs.Cpus[0].Count == 0
+
+	return !isInvalid
+}
+
+func parseMemValue(v string) (uint64, error) {
+	var scaler = uint64(1)
+	if strings.HasSuffix(v, "GB") {
+		scaler = 1024
+		v = strings.TrimSuffix(v, "GB")
+	}
+
+	val, err := strconv.ParseUint(v, 10, 64)
+	return val * scaler, err
+}
+
+func (e *environ) findInstanceSpec(controller bool, allImages []*imagemetadata.ImageMetadata, instanceTypes []instances.InstanceType, ic *instances.InstanceConstraint) (*instances.InstanceSpec, error) {
+	version, err := series.UbuntuSeriesVersion(ic.Series)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	oss, _, err := e.packetClient.OperatingSystems.List()
+	if err != nil {
+		return nil, err
+	}
+	suitableImages := []*imagemetadata.ImageMetadata{}
+
+	for _, it := range instanceTypes {
+		for _, os := range oss {
+			if os.Distro == "ubuntu" && os.Version == version {
+				for _, p := range os.ProvisionableOn {
+					if p == it.Name {
+						image := &imagemetadata.ImageMetadata{
+							Id:   os.Slug,
+							Arch: arch.AMD64,
```

Comment:
This should be populated from the metadata obtained via the packet API. Juju allows you to request (via a constraint) to bootstrap on non-amd64 instances.

Reply:
so from instanceType we create images of the same slug with different archs?